Update README.md to reflect the upcoming S4TF 0.2 release. (#114)
tensorflow/swift: README.md

# Swift for TensorFlow

> S4TF: No barriers.

> Swift for TensorFlow (aka S4TF) is a next-generation machine learning framework,

Author: How about instead: "next-generation system for machine learning'. I agree framework isn't quite right, but I also don't think language is right either. What do you think?

> Swift for TensorFlow (aka S4TF) is a next-generation machine learning framework,

Reviewer: System sounds too low-level. How about "platform"?

> Swift for TensorFlow (aka S4TF) is a next-generation machine learning framework,

Author: I like it. Going to go with that for now. We can continue to bikeshed the wording here after this PR gets merged. :-)

differential programming, systems design, and beyond. This project is
approaching version _0.2_; it is neither feature complete nor production-ready.
But it is ready for _pioneers_ to try it for your own projects, give us
feedback, and help shape the future!

The S4TF project is currently focusing on 2 kinds of users:

 1. **Advanced ML Researchers** who are limited by current ML frameworks.
    S4TF's advantages include a seamless integration with a modern general-purpose
    language, allowing for more dynamic and sophisticated models. Fast
    abstractions can be developed "in user-space" (as opposed to in C/C++
    aka "framework-space"), resulting in modular APIs that can be easily
    customized.

 2. **ML Learners** who are just getting started with machine learning. Thanks
    to Swift's support for quality tooling (e.g. context-aware autocomplete),
    S4ML can be one of the most productive ways to get started learning the
    fundamentals of machine learning.

## Getting started

 - **Colab**: The fastest way to get started is to try out S4TF right in your
   browser. Just open up our [getting started

>    notebook](https://colab.research.google.com/github/tensorflow/swift-tutorials/blob/master/iris/swift_tensorflow_tutorial.ipynb) (or start from a

Reviewer 2: The getting started notebook is exactly what I was looking for to get started myself! I don’t know enough to comment on the content, but I’ll be sure to provide feedback once I’ll go through the tutorial.

   [blank notebook](https://colab.research.google.com/github/tensorflow/swift/blob/master/notebooks/blank_swift.ipynb))!

>    Read more in our [usage guide](Usage.md).

Reviewer 2: I clicked on a some of the links in the readme, and the links that I clicked look alive and well. I wonder whether there’s some sort of automation that checks that.

 - **Install locally**: you can [download a pre-built Swift for TensorFlow
   package](Installation.md). After installation, you can follow these

>    [step-by-step instructions](Usage.md) to build and execute a S4TF program

Reviewer: ```suggestion
   [step-by-step instructions](Usage.md) to build and execute a Swift script
```

   on your computer.

 - **Compile from source**: If you'd like to customize S4TF or even contribute
   back, follow our [instructions on building the S4TF compiler from source,

>    from in the compiler's README](https://github.com/apple/swift/tree/tensorflow).

Reviewer: It's better to avoid links that are too long. In this case, the sentence is just too long and the second half feels redundant.

Just "instructions on building the compiler from source" would be good.

Please do join the
[swift@tensorflow.org mailing list](https://groups.google.com/a/tensorflow.org/d/forum/swift)
to hear the latest announcements, get help, and share your thoughts!

## Why S4TF?

Swift for TensorFlow is a new way to develop machine learning models. It
gives you the power of
[TensorFlow](https://www.tensorflow.org) directly integrated into the
[Swift programming language](https://swift.org/about). We believe that
machine learning paradigms are so important that they deserve
**first-class language and compiler support**. 

> Optimizing parameters by taking the gradient of functions is a fundamental

Reviewer: Append to the last sentence: 
", or differentiate with respect to an entire model by calling standard library method `model.gradient { ... }`."

> Optimizing parameters by taking the gradient of functions is a fundamental

Author: Fixed up. Thanks!

primitive of machine learning. With Swift for TensorFlow,
you can write a normal function using arbitrary data types, and then
take the derivative of it with just a single annotation.

```swift
import TensorFlow

var x = Tensor<Float>([[1, 2], [3, 4]])

@differentiable

> func myWackyPowPlus(ofBase base: Tensor<Float>,

Reviewer 3: I think a different example may be more suitable here.

`myWackyPowPlus` is actually not differentiable wrt the `exponent: Int` argument:
`error: can only differentiate with respect to parameters that conform to 'Differentiable', but 'Int' does not conform to 'Differentiable'`

Also, I think including an example call to a differential operator like `gradient(of:)` at the bottom of the code snippet is crucial.

---

I'll update this comment with an updated AD example in a bit!

Reviewer: I don't really like showing `@differentiable` because it doesn't show any actual differentiation. Showing the use of a differential operator (e.g. `Differentiable.gradient(in:)`) with a trailing closure is most clear.

Reviewer 3: Here are two examples demonstrating differential operators:
```swift
import TensorFlow
let 𝛁x = gradient(at: 4.0) { x in
  x * x * x
}
print(𝛁x) // 48.0
```

```swift
// Custom differentiable type.
struct Model : Differentiable {
  var w: Float
  var b: Float
  func applied(to input: Float) -> Float {
    return w * input + b
  }
}
// Differentiate using `Differentiable.gradient(at:)`.
let (𝛁model, 𝛁input) = Model(w: 4.0, b: 3.0).gradient(at: 2.0) { (model, input) in
  model.applied(to: input)
}
print(𝛁model) // Model.AllDifferentiableVariables(w: 2.0, b: 1.0)
print(𝛁input) // 4.0
```

I'd recommend using the second one on the README.

Reviewer: I like the second example, but the closure sticks out on the right hand side. Moving `Model(w: 4.0, b: 3.0)` to a temporary variable would be better.

Author: I've updated to the second one, as that's pretty nice. But what do you think about making it use Tensors instead of just floats? Or a mix of Tensors and standard floats?

Reviewer: +1 to mixing.

Reviewer: Except you'll end up with a super wacky model.

> func myWackyPowPlus(ofBase base: Tensor<Float>,

Author: Resolving this thread in favor of the other thread that @dan-zheng started below.

As a side note: can we switch the line noise in that example (`{$0}`) to just `identity`? I think that would make it much easier to read. (At first I thought I had a weird copy-paste bug!)

> func myWackyPowPlus(ofBase base: Tensor<Float>,

Reviewer: Yes, I was about to add identity to the library.

                    toExponent exponent: Int) -> Tensor<Float>
  var x = base
  for _ in 1...exponent {
      x += x • base // alternatively: x += matmul(x, base)
      print(x)  // Debugging!
  }
  return x
}
```

Beyond derivatives, the S4TF project comes with a sophisticated toolchain
to make users more productive. You can run Swift interactively in a Jupyter
notebook, and get helpful autocomplete suggestions to help you explore the
massive API surface of a modern deep learning library. You can [get started
right in your browser in
seconds](https://colab.research.google.com/github/tensorflow/swift-tutorials/blob/master/iris/swift_tensorflow_tutorial.ipynb)!

Migrating to S4TF is really easy thanks to Swift's powerful Python integration.
You can incrementally migrate your Python code over (or continue to use your
favorite Python libraries), because you can easily call your favorite Python
library with a familiar syntax:

```swift

> import Python

Author: I was alpha sorting. Is that not idiomatic?

> import Python

Reviewer: /Not/ not idiomatic at all. Was thinking of putting TF first, but it's not important.

> import Python

Author: Okay. Fixed to TF then Python.

import TensorFlow

> 

Author: Sorry, not sure what you mean for this? Does this need to be in the front-page README?

> 

Reviewer: I meant showing `PythonLibrary.useVersion(3, 7)` after you import Python to show that you can use any version of Python you have.

> 

Author: Hmm, I think that's too much unnecessary detail for the top-level README. Does that make sense? 

That said, we should definitely have that documented somewhere easily accessible.

> 

Reviewer: SGTM. I don't have a strong opinion about it at all.

let np = Python.import("numpy")

> let myNumpyArray = np.arange([10, 10])  // Create a 10x10 numpy array.

Reviewer: Yes, and I'd choose `array` since the code is short enough.

- [Swift Tensor API Reference](https://www.tensorflow.org/api_docs/swift/Structs/Tensor)

### Technology deep dive

The Swift for TensorFlow project builds on top of powerful theoretical
foundations. For insight into some of the underlying technologies, check
out the following documentation.

> Beware: the project is moving very quickly, and thus some of these documents
> are slightly out of date as compared to the current state-of-the-art.

- [Automatic Differentiation Whitepaper](docs/AutomaticDifferentiation.md)
- [Automatic Differentiation Manifesto](https://gist.github.com/rxwei/30ba75ce092ab3b0dce4bde1fc2c9f1d)
- [Python Interoperability](docs/PythonInteroperability.md)
- [Aggregate Parameters and Parameter Update](docs/AggregateParameters.md)
- [Graph Program Extraction](docs/GraphProgramExtraction.md)

## Source code

Compiler and standard library development of Swift for TensorFlow happens
on the `tensorflow` branch of the
[apple/swift](https://github.com/apple/swift/tree/tensorflow) repository.

Additional code repositories that make up the core of project include:

 - [Debugger and REPL support](http://github.com/apple/swift-lldb/tree/tensorflow)

>  - [High level APIs](https://github.com/tensorflow/swift-apis)

Reviewer: We've been calling it the "deep learning library" instead of "high-level APIs". "High level" is too heavily overloaded.

Author: Yeah, that's much better. I've also switched LLDB to match the new pattern. Thanks!

> Swift for TensorFlow is **not** intended to remain a long-term fork of the official 
> Swift language. New language features will eventually go through the Swift evolution process
> as part of being considered for being pulled into master.

### Jupyter Notebook support

[Jupyter Notebook](http://jupyter.org/) support for Swift is under development at
[google/swift-jupyter](https://github.com/google/swift-jupyter).

## Community

Swift for TensorFlow discussions happen on the
[swift@tensorflow.org mailing list](https://groups.google.com/a/tensorflow.org/d/forum/swift).

For the most part, the core team's development is tracked in

> [JIRA].

Reviewer 2: Looks like this sentence is a duplicate of what comes in the next section?

Author: Great catch @xeno-by ! It's tinkercruft leftover from when I was working through it. (Hence the broken link.) It should now be removed.

### Bugs Reports and Feature Requests

Before reporting an issue, please check the [Frequently Asked Questions](FAQ.md)
to see if your question has already been addressed.

For questions about general use or feature requests, please send an email to
the [mailing list](mailto:swift@tensorflow.org) or search for relevant issues
in the [JIRA issue tracker](https://bugs.swift.org/projects/TF/issues/?filter=allopenissues).

For the most part, the core team's development is also tracked in
[JIRA](https://bugs.swift.org/secure/RapidBoard.jspa?rapidView=17&projectKey=TF&view=planning).

### Contributing

We welcome source code contributions: please read 
[Contributing Code](https://swift.org/contributing/#contributing-code).
It is always a good idea to discuss your plans on the mailing list before
making any major submissions.

Check out our [starter bugs](https://bugs.swift.org/issues/?filter=11323) if

> you'd like to get involved but don't know where to start!

Reviewer 2: It’s very nice that there’s a bunch of starter bugs to choose from!

Author: 👍

# Swift for TensorFlow

> Swift for TensorFlow: No barriers.

> Swift for TensorFlow (aka S4TF) is a next-generation platform for machine

Reviewer: I think this "S4TF" should be removed as well for more clear, consistent branding.

Author: Done!

- [Sample Models](https://github.com/tensorflow/swift-models)
- [Tutorials](https://github.com/tensorflow/swift-tutorials)
- [Frequently Asked Questions](FAQ.md)
- [Swift Tensor API Reference](https://www.tensorflow.org/api_docs/swift/Structs/Tensor)

### Technology deep dive

The Swift for TensorFlow project builds on top of powerful theoretical
foundations. For insight into some of the underlying technologies, check
out the following documentation.

> Beware: the project is moving very quickly, and thus some of these documents
> are slightly out of date as compared to the current state-of-the-art.

- [Automatic Differentiation Whitepaper](docs/AutomaticDifferentiation.md)
- [Automatic Differentiation Manifesto](https://gist.github.com/rxwei/30ba75ce092ab3b0dce4bde1fc2c9f1d)
- [Python Interoperability](docs/PythonInteroperability.md)
- [Aggregate Parameters and Parameter Update](docs/AggregateParameters.md)
- [Graph Program Extraction](docs/GraphProgramExtraction.md)

## Source code

Compiler and standard library development happens on the `tensorflow` branch of
the [apple/swift](https://github.com/apple/swift/tree/tensorflow) repository.

> Additional code repositories that make up the core of project include:

Reviewer 4: Small typo I found. Excited about the progress! 😄 
```suggestion
Additional code repositories that make up the core of the project include:
```

Author: Thanks for the catch! Fixed up.

   browser. Just open up our [getting started
   notebook](https://colab.research.google.com/github/tensorflow/swift-tutorials/blob/master/iris/swift_tensorflow_tutorial.ipynb) (or start from a
   [blank notebook](https://colab.research.google.com/github/tensorflow/swift/blob/master/notebooks/blank_swift.ipynb))!
   Read more in our [usage guide](Usage.md).

 - **Install locally**: you can [download a pre-built Swift for TensorFlow
   package](Installation.md). After installation, you can follow these
   [step-by-step instructions](Usage.md) to build and execute a Swift script
   on your computer.

 - **Compile from source**: If you'd like to customize S4TF or even contribute
   back, follow our [instructions on building the S4TF compiler from
   source](https://github.com/apple/swift/tree/tensorflow).

Please do join the
[swift@tensorflow.org mailing list](https://groups.google.com/a/tensorflow.org/d/forum/swift)
to hear the latest announcements, get help, and share your thoughts!

## Why Swift for TensorFlow?

Swift for TensorFlow is a new way to develop machine learning models. It
gives you the power of

> [TensorFlow](https://www.tensorflow.org) directly integrated into the

Reviewer: I think we should mention "protocol-oriented differentiable programming" since the vision here is bigger than ML and TensorFlow.
Maybe here: "the power of `TensorFlow` and protocol-oriented differentiable programming"
But I'm not sure.

Author: Hmm, I think this is becoming quite wordy, and trying to wedge that in seems a little unnecessary given that the next paragraph and code demonstration cover the concept pretty effectively. Does that make sense?

Reviewer: SGTM!

[Swift programming language](https://swift.org/about). We believe that
machine learning paradigms are so important that they deserve
**first-class language and compiler support**. 

A fundamental primitive in machine learning is gradient-based optimization:
computing function derivatives to optimize parameters. With Swift for
TensorFlow, you can easily differentiate custom functions using differential
operators like `gradient(of:)`, or differentiate with respect to an entire
model by calling standard library method `model.gradient { ... }`.

```swift
// Custom differentiable type.

> struct Model : Differentiable {

Reviewer: In public code, we use 4-space indentation and the "no space before any colon" style.

```suggestion
struct Model: Differentiable {
```

Author: Ack; fixed.

  var w: Float
  var b: Float
  func applied(to input: Float) -> Float {
    return w * input + b
  }
}

> // Differentiate using `Differentiable.gradient(at:)`.

Reviewer: ```suggestion
// Differentiate using `Differentiable.gradient(at:in:)`.
```

Author: Fixed up. Thanks!

let model = Model(w: 4.0, b: 3.0)

> let (𝛁model, 𝛁input) = model.gradient(at: 2.0) { (model, input) in

Reviewer: ```suggestion
let (𝛁model, 𝛁input) = model.gradient(at: 2.0) { model, input in
```

Author: Fixed.

```

## Documentation

### Getting started

- [Swift for TensorFlow Design Overview](docs/DesignOverview.md)
- [Why *Swift* for TensorFlow?](docs/WhySwiftForTensorFlow.md)
- [Sample Models](https://github.com/tensorflow/swift-models)
- [Tutorials](https://github.com/tensorflow/swift-tutorials)]
- [Frequently Asked Questions](FAQ.md)
- [Swift Tensor API Reference](https://www.tensorflow.org/api_docs/swift/Structs/Tensor)

### Technology deep dive

The Swift for TensorFlow project builds on top of powerful theoretical
foundations. For insight into some of the underlying technologies, check
out the following documentation.

> Beware: the project is moving very quickly, and thus some of these documents
> are slightly out of date as compared to the current state-of-the-art.

- [Automatic Differentiation Whitepaper](docs/AutomaticDifferentiation.md)
- [Automatic Differentiation Manifesto](https://gist.github.com/rxwei/30ba75ce092ab3b0dce4bde1fc2c9f1d)
- [Python Interoperability](docs/PythonInteroperability.md)
- [Aggregate Parameters and Parameter Update](docs/AggregateParameters.md)
- [Graph Program Extraction](docs/GraphProgramExtraction.md)

## Source code

> Compiler and standard library development of Swift for TensorFlow happens

Reviewer: Remove redundant words?
```suggestion
Compiler and standard library development happens
```

Author: Yup. Good catch. Fixed.

on the `tensorflow` branch of the
[apple/swift](https://github.com/apple/swift/tree/tensorflow) repository.

Additional code repositories that make up the core of project include:

 - [Swift fork of LLDB](http://github.com/apple/swift-lldb/tree/tensorflow):
   debugger and REPL support.

>  - [Deep Learning library](https://github.com/tensorflow/swift-apis): a high-level

Reviewer: ```suggestion
 - [Deep learning library](https://github.com/tensorflow/swift-apis): a high-level
```

Author: Fixed up.

>  - [Deep Learning library](https://github.com/tensorflow/swift-apis): a high-level

Reviewer: "a high-level API" -> "high-level APIs"

Author: Fixed up. Thanks!
